src/librustc_typeck/collect/type_of.rs

```diff
@@ -256,15 +256,18 @@ pub(super) fn type_of(tcx: TyCtxt<'_>, def_id: DefId) -> Ty<'_> {
                         // figure out which generic parameter it corresponds to and return
                         // the relevant type.
                         let generics = match path.res {
-                            Res::Def(DefKind::Ctor(..), def_id) => {
-                                tcx.generics_of(tcx.parent(def_id).unwrap())
-                            }
+                            Res::Def(
+                                DefKind::Ctor(..) | DefKind::AssocTy | DefKind::AssocFn,
```

Comment:
Isn't `AssocFn` impossible to hit here? Oh, except maybe for trait methods?
But even for trait methods, won't you miss the segment of the method itself?

Reply:
```
#![feature(const_generics)]
//~^ WARN the feature `const_generics` is incomplete and may cause the compiler to crash

trait T<const A: usize> {
    fn f();
}
struct S;

impl T<0usize> for S {
    fn f() {}
}

fn main() {
    let _err = <S as T<0usize>>::f();
    // ^ this hits AssocFn
}
```

Reply:
That's what I meant by trait methods, yes.